### PR TITLE
Support Google Cloud Storage in addition to S3

### LINKED ISSUE #1566: Support Google Cloud Storage in addition to S3
Normandy admin currently uploads addons to an S3 bucket. Normandy should support uploading to a GCS bucket as well. 

Currently we have an DJANGO_AWS_STORAGE_BUCKET_NAME bucket name. Maybe we could also have a DJANGO_GCS_STORAGE_BUCKET_NAME, and the one that is set determines which storage to use?

### PR DESCRIPTION
**Pull request by mythmon**
Fixes #1566

### CODE REVIEW

File: normandy/base/storage.py [hunk 1]
```diff
@@ -30,9 +31,10 @@ def get_valid_name(self, name):
         """
 
         name = force_text(name).strip()
-        # remove "characters to avoid", as described by S3's docs
+        # remove "characters to avoid", as described by S3's GCP's docs
```
Comment (sciurus) [inline normandy/base/storage.py]: Nit: _S3's and GCP's_

File: normandy/base/storage.py [hunk 2]
```diff
@@ -30,9 +31,10 @@ def get_valid_name(self, name):
...
         # https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingMetadata.html#object-key-guidelines-avoid-characters
-        name = re.sub(r'[\\{}^%`\[\]<>~#|\x00-\x1F\x7F-\xFF\'"]', "", name)
+        # https://cloud.google.com/storage/docs/naming#objectnames
+        name = re.sub(r'[\\{}^%`\[\]<>~#|\x00-\x1F\x7F-\xFF\'"*?]', "", name)
```
Comment (sciurus) [inline normandy/base/storage.py]: Our of curiosity and paranoia, I manually tested this successfully.

**Comment by peterbe** [comment]
Can you merge master? @mythmon

File: normandy/base/storage.py [hunk 3]
```diff
@@ -6,20 +6,21 @@
...
 from inmemorystorage import InMemoryStorage
 from storages.backends.s3boto3 import S3Boto3Storage
+from storages.backends.gcloud import GoogleCloudStorage
 
 
 class PermissiveFilenameStorageMixin(object):
     """
     A storage class mixin that is is more permissive about valid filenames.
 
     In the original implemtnation, "anything that is not a unicode
```
Comment (peterbe) [inline normandy/base/storage.py]: might as well fix this "implemtnation" typo :)

File: normandy/base/storage.py [hunk 4]
```diff
@@ -6,20 +6,21 @@
...
-    converted to underscores.
+    alphanumeric, dash, underscore, or dot is removed". This modified version
+    allows most characters, except a few special characters. Leading and
+    trailing whitespaces are removed and internal whitespace is converted to
+    underscores.
 
-    Characters that S3 describes as requiring "significant special
-    handling for consistency across all applications" are removed.
+    File names produced follow recommended guidelines for both AWS S3 and GCP
+    Cloud Storage object names.
```
Comment (peterbe) [inline normandy/base/storage.py]: nit: If you have a URL to that "recommended guidelines" that would be great.

Comment (mythmon) [reply normandy/base/storage.py]: The URL is linked in the code, next to the actual changes to follow the guidelines. Does that work?

**Comment by mythmon** [comment]
I fixed the nits and rebased on master to fix merge conflicts.

> I have not tested this on a real GCS account, have you?

I have not tested this directly either. I haven't worked on that kind of set up yet. I'm comfortable merging this without that testing though, since we aren't deploying anything just yet.

**Comment by peterbe** [comment]
I've actually never tested any real S3 uploads from localhost myself. That's how far behind I am :)

Next up on my personal todo list is to get comfortable with GCS for the sake of Tecken so hopefully I can leverage some of that experience in normandy/DC and test for realz.

**Comment by mythmon** [comment]
Nits are fixed, and I have approvals. Removing the label and merging this.

bors r= @peterbe, @sciurus

**Pull request merged**